homeassistant/components/homepluscontrol/switch.py

```diff
@@ -0,0 +1,216 @@
+    DEVICE_CLASS_OUTLET,
+    DEVICE_CLASS_SWITCH,
+    SwitchEntity,
+)
+
+# from homeassistant.helpers import entity_platform
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import DOMAIN, HW_TYPE
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Legrand Home+ Control Switch platform in HomeAssistant.
+
+    Args:
+        hass (HomeAssistant): HomeAssistant core object.
+        config_entry (ConfigEntry): ConfigEntry object that configures this platform.
+        async_add_entities (function): Function called to add entities of this platform.
+    """
+    # Dictionaty of entities of this integration
+    hass.data[DOMAIN]["entities"] = {}
+
+    # API object stored here by __init__.py
+    api = hass.data[DOMAIN][config_entry.entry_id]
+
+    async def async_update_data():
+        """Fetch data from API endpoint.
+
+        This is the place to pre-process the data to lookup tables
+        so entities can quickly look up their data.
+        """
+        switch_data = {}
+        try:
+            # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+            # handled by the data update coordinator.
+            async with async_timeout.timeout(10):
+                switch_data = await api.fetch_data()
+        except Exception as err:
+            raise UpdateFailed(
+                f"Error communicating with API: {err} [{type(err)}]"
+            ) from err
+
+        # Remove obsolete entities from the Entity and Device Registries
+        device_reg = await hass.helpers.device_registry.async_get_registry()
+
+        for ent_id, ent in api.switches_to_remove.items():
+            r_entity = hass.data[DOMAIN]["entities"].pop(ent_id, None)
+            if r_entity is None:
+                _LOGGER.debug(
+                    "Entity %s was marked from deletion, but is not registered by the integration.",
+                    str(ent),
+                )
+                continue
+            _LOGGER.debug(
+                "Remove entity %s [device: %s] from HA registries.",
+                r_entity.entity_id,
+                r_entity.registry_entry.device_id,
+            )
+            device_reg.async_remove_device(r_entity.registry_entry.device_id)
+            await r_entity.async_remove()
+
+        # Reset the api object dictionary of deleted elements
+        api.switches_to_remove = {}
+
+        # Add entities to HomeAssistant domain and registries
+        new_entities = []
+        for idx in switch_data:
+            if idx not in hass.data[DOMAIN]["entities"]:
+                hass.data[DOMAIN]["entities"][idx] = HomeControlSwitchEntity(
+                    coordinator, idx
+                )
+                new_entities.append(hass.data[DOMAIN]["entities"][idx])
+
+        async_add_entities(new_entities)
```

Comment:
Consider separating the entity addition from the coordinator and instead signal the entity addition from the coordinator via our dispatch helper. That will allow the same coordinator to handle multiple platforms which may be needed in the future. It also separates concerns better.

Reply:
I introduced signaling functionality here through the dispatchers so that entities are removed and added through specific signals that are sent out. Thanks for the suggestion.